test/schedulers.py

```diff
@@ -0,0 +1,50 @@
+import test_utils
+from torch.optim.lr_scheduler import _LRScheduler
+
+MIN_DELTA_TO_UPDATE_LR = 0.000001
+NUM_WARMUP_EPOCHS = 1
+
+
+class WarmupAndExponentialDecayScheduler(_LRScheduler):
+  """TODO(zcain): docstring."""
+  def __init__(self, optimizer, num_steps_per_epoch, divide_every_n_epochs=20, divisor=5):
+    self.num_steps_per_epoch = num_steps_per_epoch
+    self.divide_every_n_epochs = divide_every_n_epochs
+    self.divisor = divisor
+    self.previous_lr = -1
+    self.max_lr = optimizer.param_groups[0]['lr']
+    self.optimizer = optimizer
+
+  def get_lr(self, epoch, step):
```

Comment:
Following the PyTorch `_LRScheduler` [API](https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html) should we do `def get_lr(self)` and instead track the global step count or epoch&step as member variables and increment it at `self.step()` call time? What do you think?

Reply:
I wasn't sure: does each core share the context objects, e.g. optimizer? Or is there a separate optimizer for every core? https://github.com/pytorch/xla/blob/master/test/test_train_imagenet.py#L147

For 8 cores, we'll make 8 calls to scheduler.step(). If these are all separate scheduler objects, it's fine to do what you suggested. If they share the object, then we wouldn't want to save state like this

Reply:
The scheduler takes an optimizer, which is per device.
Hence the scheduler must be per device.
You cannot manipulate tensors which are on device X from loop function handling device Y.

Reply:
https://github.com/pytorch/xla/blob/master/torch_xla_py/data_parallel.py#L195

Reply:
Thanks for the explanation. In this case, I think it makes sense to track step and epoch internally. Made that switch and now the scheduler.step() call in the train loop is very simple